src/plat/qemu-arm-virt/config.cmake

```diff
@@ -34,9 +37,32 @@ endmacro()
 
 if(KernelPlatformQEMUArmVirt)
 
+    # Our default QEMU configuration is AARCH64/Cortex-a53. If KernelSel4Arch is
+    # set to aarch32, the default is Cortex-a15. Setting ARM_CPU allows picking
+    # a specific CPU and forces KernelSel4Arch to the respective architecture.
+    # Note that the QEMU virt platform has some limitations:
+    # - only specific ARMv7 and ARMv8 cores can be used.
+    # - ARMv8 cores can't be started in AARCH32 mode
+    # - Cortex-A32 that implements AARCH32 only is not suppoted
+
     if(NOT ARM_CPU)
-        message(STATUS "ARM_CPU not set, defaulting to cortex-a53")
-        set(ARM_CPU "cortex-a53")
+        # We don't have any KernelSel4Archxxx varable set up here, because
+        # declare_seL4_arch() has no been called yet. So we must check
```

Comment:
The "yet" implies it will be called in the future, but that isn't the case, is it? Because if it was, we shouldn't guess something here, but wait for that future call instead.

This comment doesn't explain why `setup_qemu_armv7` should set anything. I think the intention is to have sane defaults in case nothing was specified, but that intention isn't clear from the comments.

Reply:
CMake actually iterates over the configuration multiple times, until it becomes stable (or abort after 100 iterations if not) . Also, CMake cache some configuration. So what one will see that the variable can be set in some cases when we pass here. Only when you do a clean build, one will see that it is not set n the first iteration). 

`setup_qemu_armv7()`/`setup_qemu_armv7()` will call `declare_seL4_arch()` based on the selected CPU in the code below. We have to solve a check-egg problem here. In the seL4  build system `declare_seL4_arch()` sets the default and understands KernelSel4Arch`. But it does not know about QEMU specific `ARM_CPU` that one can specify. So we have to do some guessing here first t ensure `declare_seL4_arch()` picks the right default then.

Reply:
My worry is that we're working around CMake file bugs in projects here instead of fixing it there. Guessing here will make it much harder to fix problems in the future, even though we may have less problems initially.

Reply:
I think what we have now is the proper way to handle this. There is not much so guessing involved actually, the rulsed are quite simple:

| Config | Result | Notes
|--------|--------|--------|
| set `ARM_CPU` | use it (or error) and overwrite `KernelSel4Arch` | raise error if both `ARM_CPU` and `KernelSel4Arch` are set to conflicting or unsupported combinations?  |
| set `KernelSel4Arch` | cortex-A15 or cortex-A53 (or error) |
| nothing | cortex-A53 | this case was broken as described in https://github.com/seL4/seL4/pull/824#issuecomment-1921221807. Seems ARMv7 QEMU is rarely used nowadays, thus we did not see this earlier |
| conflicting | raises an error | fix project, e.g.  https://github.com/seL4/sel4webserver/pull/18 |

Now, if a projects change the variables after the seL4 build system CMake files have run, things can get tricky. Since CMake iterates over the config, it may end up in a stable state. But we may also have broken some older project that do this. I'd leave it to somebody raising the issues then, so we can check what the project does. If it's not somewhere in CI or somebody actively using it, things might also be obsolete eventually.

Reply:
Silently overwriting `KernelSel4Arch` is what seems wrong to me. But like you said, it may be necessary in case of reconfiguration, if config values are cached.

Reply:
I will add some check that this fail the build then if both are set and conflicting.

Reply:
Turns out that this case is handles already, as `KernelSel4Arch` is not overwritten anywhere. It is set in case the user has not set it. And if the user has set it, then the conflict will be notices already. So things should be fine here also.